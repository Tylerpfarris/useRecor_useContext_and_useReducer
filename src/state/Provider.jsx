import React, { createContext, useContext, useReducer } from 'react';

const useRecordContext = createContext();


export const Provider = ({ children, reducer, initialState }) => {
   const [state, dispatch] = useReducer(reducer, initialState);
   return (
      <useRecordContext.Provider value={{ state, dispatch }}>{children}</useRecordContext.Provider>
   );
};

export const useSelector = (selector) => {
   const { state } = useContext(useRecordContext);
   return selector(state);
}

export const useDispatch = () => {
   const { dispatch } = useContext(useRecordContext);
   return dispatch;
}

