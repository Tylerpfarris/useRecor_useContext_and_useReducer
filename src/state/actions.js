export const DO_UNDO = 'DO_UNDO';
export const doUndo = () => ({
   type: DO_UNDO,
})
export const DO_REDO = 'DO_REDO';
export const doRedo = () => ({
   type: DO_REDO,
})
export const DO_RECORD = 'DO_RECORD';
export const doRecord = (init) => ({
   type: DO_RECORD,
   payload: init
})