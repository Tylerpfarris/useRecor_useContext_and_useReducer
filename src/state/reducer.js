export const initialState = {
   before: [],
   after: [],
   current: '#FF0000',
};

export function recordReducer(state, action) {
   switch (action.type) {
      case 'DO_UNDO':
         return {
            ...state,
            before: state.before.slice(0, -1),
            after: [state.current, ...state.after],
            current: state.before[state.before.length - 1]
         };
      case 'DO_REDO':
         return {
            ...state,
            before: [...state.before, state.current],
            after: state.after.slice(1),
            current: state.after[0]
         };
      case 'DO_RECORD':
         return {
            ...state,
            before: [...state.before, state.current],
            current: action.payload
         };
      default:
         return state;
   }
}