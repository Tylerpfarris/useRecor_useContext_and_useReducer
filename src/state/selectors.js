export const selectAllState = (state) => state;
export const doBefore = (state) => state.before;
export const doAfter = (state) => state.after;
export const doCurrent = (state) => state.current;
