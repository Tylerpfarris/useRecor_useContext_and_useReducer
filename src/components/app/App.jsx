import React from 'react';
import { doUndo, doRedo, doRecord } from '../../state/actions';
import { doCurrent } from '../../state/selectors';
import { useDispatch, useSelector } from '../../state/Provider';



function App() {
  const dispatch = useDispatch();
  
  const current = useSelector(doCurrent);
  return (
    <>
      <button data-testid="undo-button" onClick={() => dispatch(doUndo())}>
        undo
      </button>
      <button data-testid="redo-button" onClick={() => dispatch(doRedo())}>
        redo
      </button>
      <input
        data-testid='color-selector'
        type="color"
        value={current}
        onChange={({ target }) => dispatch(doRecord(target.value))}
      />
      <div
        data-testid='display-div'
        style={{ backgroundColor: current, width: '10rem', height: '10rem' }}
      ></div>
    </>
  );
}

export default App;
