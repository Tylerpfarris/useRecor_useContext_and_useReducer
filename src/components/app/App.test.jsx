import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';
import { Provider } from '../../state/Provider';
import { initialState, recordReducer } from '../../state/reducer';

describe('App component', () => {
  afterEach(() => cleanup());
  it('renders App', () => {
    const { asFragment } = render(
      <Provider
        reducer={recordReducer}
        initialState={initialState}
      >
        <App />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();

    const undoEl = screen.getByTestId('undo-button');
    expect(undoEl).toHaveTextContent('undo');
    const redoEl = screen.getByTestId('redo-button');
    expect(redoEl).toHaveTextContent('redo');
    const displayEl = screen.getByTestId('display-div');
    expect(displayEl).toHaveStyle({
          backgroundColor: 'current',
          width: '10rem',
          height: '10rem',
    });
    const colorSelectorEl = screen.getByTestId('color-selector');
    expect(colorSelectorEl).toHaveAttribute('type', 'color');

        fireEvent.input(colorSelectorEl, { target: { value: '#333333' } });
        expect(colorSelectorEl).toHaveValue('#333333');

        fireEvent.input(colorSelectorEl, { target: { value: '#ff3333' } });
        expect(colorSelectorEl).toHaveValue('#ff3333');

        userEvent.click(undoEl);
        expect(colorSelectorEl).toHaveValue('#333333');

        userEvent.click(redoEl);
        expect(colorSelectorEl).toHaveValue('#ff3333');
  });
});
