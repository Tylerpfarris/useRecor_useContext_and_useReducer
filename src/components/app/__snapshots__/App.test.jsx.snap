// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders App 1`] = `
<DocumentFragment>
  <button
    data-testid="undo-button"
  >
    undo
  </button>
  <button
    data-testid="redo-button"
  >
    redo
  </button>
  <input
    data-testid="color-selector"
    type="color"
    value="#FF0000"
  />
  <div
    data-testid="display-div"
    style="background-color: rgb(255, 0, 0); width: 10rem; height: 10rem;"
  />
</DocumentFragment>
`;
